include/sys/zfs_ioctl.h

```diff
@@ -160,16 +160,41 @@ typedef enum dmu_send_resume_token_version {
  * cannot necessarily be received as a clone correctly.
  */
 #define	DRR_FLAG_FREERECORDS	(1<<2)
+/*
+ * When DRR_FLAG_SPILL_BLOCK is set it indicates the DRR_SPILL_BLOCK flag
+ * is meaningful in the send stream.
+ *
+ * When DRR_FLAG_SPILL_BLOCK is set, DRR_OBJECT records will have
+ * DRR_SPILL_BLOCK set if and only if they should have a spill block
+ * (either an existing one, or a new one in the send stream).  When clear
+ * the object does not have a spill block and any existing spill block
+ * should be freed.
+ *
+ * Similarly, when DRR_FLAG_SPILL_BLOCK is set, DRR_SPILL records will
+ * have DRR_SPILL_BLOCK set if and only if they were included for backward
+ * compatibility purposes, and can be safely ignored by new versions of
+ * zfs receive.  Previous versions of ZFS which do not understand the
+ * DRR_FLAG_SPILL_BLOCK will process this record and recreate any missing
+ * spill blocks.
+ *
+ * DRR_SPILL_BLOCK has different meanings in different record types
+ * because the meanings are closely linked, there is per-record-type
+ * specific behavior, and there is only a small amount of space (8 bits)
+ * in drr_flags in the drr_object and drr_spill structures.
+ */
+#define	DRR_FLAG_SPILL_BLOCK	(1<<3)
 
 /*
  * flags in the drr_flags field in the DRR_WRITE, DRR_SPILL, DRR_OBJECT,
  * DRR_WRITE_BYREF, and DRR_OBJECT_RANGE blocks
  */
 #define	DRR_CHECKSUM_DEDUP	(1<<0) /* not used for DRR_SPILL blocks */
 #define	DRR_RAW_BYTESWAP	(1<<1)
+#define	DRR_SPILL_BLOCK		(1<<2)
```

Comment:
The name of this macro is not that descriptive.  What would you think about adding 2 macros, one for each use:
```
DRR_OBJECT_HAS_SPILL 1<<2 // this OBJECT record has a spill block
DRR_SPILL_IS_UNMODIFIED 1<<2  // this SPILL record describes a block that is not actually modified, and can be ignored by software that proceses DRR_OBJECT_HAS_SPILL
```

Reply:
Sounds good, I opted to name the flags `DRR_OBJECT_SPILL` and `DRR_SPILL_UNMODIFIED`, and check macros `DRR_OBJECT_HAS_SPILL()` and `DRR_SPILL_IS_UNMODIFIED`.  I think this strikes a good balance between descriptive names while remaining  consistent with the existing naming scheme.  For example, only using `IS` or `HAS` in the macros which check the flag.